[NBS Docs] Update integrate search into plugins docs (#26275)
backstage/backstage: docs/plugins/integrating-search-into-plugins.md

> You can always look at a working example, e.g. [StackOverflowQuestionsCollatorFactory](https://github.com/backstage/backstage/blob/master/plugins/search-backend-module-stack-overflow-collator/src/collators/StackOverflowQuestionsCollatorFactory.ts), if you are unsure or want to follow best practices.

#### 1. Create a collator module package

> In order to add a FAQ collator to the Backstage index registry, we have to create a [plugin module](https://backstage.io/docs/backend-system/building-plugins-and-modules/index#modules) and the best way to do this is to create it in a separate package, e.g., `plugins/search-backend-module-faq-snippets-collator`, using the `backstage-cli new` command:

Reviewer: ```suggestion
In order to add a FAQ collator to the Backstage index registry, we have to create a [plugin module](https://backstage.io/docs/backend-system/building-plugins-and-modules/index#modules) and the best way to do this is to create it in a separate package, e.g., `plugins/search-backend-module-faq-snippets-collator`, using the `yarn new` command:
```

> 1. Access your Backstage project root directory and run `npx @backstage/cli new`;

Reviewer: ```suggestion
1. Access your Backstage project root directory and run `yarn new`;
```

2. When asked about what do you want to create, please select `backend-module` and hit enter;
3. Inform `search` as the plugin id and `faq-snippets-collator` as module id;
4. A `search-backend-module-faq-snippets-collator` folder should have been created in your project's "plugins" directory.

#### 2. Install the collator dependencies

We will use some libraries in the module, so let's add them to your plugin module dependencies:

```sh
# Create a new branch using Git command-line
git checkout -b tutorials/new-faq-snippets-collator

# Install the package containing the interface

> yarn workspace backstage-plugin-search-backend-module-faq-snippets-collator add cross-fetch @backstage/config @backstage/plugin-search-common @backstage/plugin-search-backend-node

Reviewer: ```suggestion
yarn workspace backstage-plugin-search-backend-module-faq-snippets-collator add node-fetch @backstage/config @backstage/plugin-search-common @backstage/plugin-search-backend-node
```

Just thinking of what we decided in the ADR: https://backstage.io/docs/architecture-decisions/adrs-adr013

}
```

Below we provide an example implementation of how the FAQ collator factory could look like using our new document type, placed in the `plugins/search-backend-module-faq-snippets-collator/src/factory.ts` file:

```ts

> import fetch from 'cross-fetch';

Reviewer: ```suggestion
import fetch from 'node-fetch';
```

import { Readable } from 'stream';

import { Config } from '@backstage/config';
import { LoggerService } from '@backstage/backend-plugin-api';
import { DocumentCollatorFactory } from '@backstage/plugin-search-common';

import { FaqSnippetDocument } from './types';

const DEFAULT_BASE_URL = 'https://backstage.example.biz/faq-snippets';

export class FaqSnippetsCollatorFactory implements DocumentCollatorFactory {
  public readonly type: string = 'faq-snippets';
  private readonly baseUrl: string;
  private readonly logger: LoggerService;

  private constructor(options: { logger: LoggerService; baseUrl: string }) {
    this.baseUrl = options.baseUrl;
    this.logger = options.logger;
  }

  static fromConfig(

>     config: Config,

Author: Fixed

            )
          : defaultSchedule;

        indexRegistry.addCollator({
          schedule: scheduler.createScheduledTaskRunner(schedule),
          factory: FaqSnippetsCollatorFactory.fromConfig(config, { logger }),
        });
      },
    });
  },
});
```

In the fragment above, the module is registered, and when the Backstage backend initializes it, it adds the FAQ Snippets collator to the search index registry. Now let's export the module as default from the `index.ts` file:

```ts title='plugins/search-backend-module-faq-snippets-collator/src/index.ts'
export { export const searchFaqSnippetsCollatorModule as default } from './module';
```

#### 7. Install the collator module

The newly created module should be added to the backend package dependencies as follows:

```sh
yarn --cwd backend add backstage-plugin-search-backend-module-faq-snippets-collator
```

After that, install the module on your Backstage backend instance:

```ts title='packages/backend/src/index.ts'
import { createBackend } from '@backstage/backend-defaults';
//...
const backend = createBackend();
// Installing the search backend plugin
backend.add(import('@backstage/plugin-search-backend/alpha'));

> // Installing the newrly created faq snippets collator module

Reviewer: ```suggestion
// Installing the newly created faq snippets collator module
```

    env.registerInit({
      deps: {
        config: coreServices.rootConfig,
        logger: coreServices.logger,
        scheduler: coreServices.scheduler,
        indexRegistry: searchIndexRegistryExtensionPoint,
      },
      async init({ config, logger, scheduler, indexRegistry }) {
        const defaultSchedule = {
          frequency: { minutes: 10 },
          timeout: { minutes: 15 },
          initialDelay: { seconds: 3 },
        };

        const schedule = config.has('faq.schedule')
          ? readSchedulerServiceTaskScheduleDefinitionFromConfig(
              config.getConfig('faq.schedule'),
            )
          : defaultSchedule;

        indexRegistry.addCollator({
          schedule: scheduler.createScheduledTaskRunner(schedule),
          factory: FaqSnippetsCollatorFactory.fromConfig(config, { logger }),
        });
      },
    });
  },
});
```

In the fragment above, the module is registered, and when the Backstage backend initializes it, it adds the FAQ Snippets collator to the search index registry. Now let's export the module as default from the `index.ts` file:

```ts title='plugins/search-backend-module-faq-snippets-collator/src/index.ts'

> export { export const searchFaqSnippetsCollatorModule as default } from './module';

Reviewer 2: ```suggestion
export { searchFaqSnippetsCollatorModule as default } from './module';
```

```

#### 7. Install the collator module

The newly created module should be added to the backend package dependencies as follows:

```sh

> yarn --cwd backend add @internal/backstage-plugin-search-backend-module-faq-snippets-collator

Reviewer 2: ```suggestion
yarn --cwd packages/backend add @internal/backstage-plugin-search-backend-module-faq-snippets-collator
```

right?

```

After that, install the module on your Backstage backend instance:

```ts title='packages/backend/src/index.ts'
import { createBackend } from '@backstage/backend-defaults';
//...
const backend = createBackend();
// Installing the search backend plugin
backend.add(import('@backstage/plugin-search-backend/alpha'));
// Installing the newly created faq snippets collator module
backend.add(
  import(
    '@internal/backstage-plugin-search-backend-module-faq-snippets-collator'
  ),
);
//...
backend.start();
```

#### 8. Testing the collator code

To verify your implementation works as expected make sure to add tests for it. For your convenience, there is the [`TestPipeline`](https://backstage.io/docs/reference/plugin-search-backend-node.testpipeline) utility that emulates a pipeline into which you can integrate your custom collator.

Look at [DefaultTechDocsCollatorFactory test](https://github.com/backstage/backstage/blob/de294ce5c410c9eb56da6870a1fab795268f60e3/plugins/techdocs-backend/src/search/DefaultTechDocsCollatorFactory.test.ts), for an example.

> You can also check out the documentation on [how to test Backstage plugin modules](https://backstage.io/docs/backend-system/building-plugins-and-modules/testing).

Reviewer 2: should this be a relative path instead of a link, like the others?

Author: Good catch, fixed 🙏🏻
